controller/src/main/java/io/pravega/controller/store/stream/PravegaTablesStreamMetadataStore.java

```diff
@@ -0,0 +1,235 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.controller.store.stream;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+import io.pravega.client.stream.StreamConfiguration;
+import io.pravega.common.Exceptions;
+import io.pravega.common.concurrent.Futures;
+import io.pravega.common.lang.Int96;
+import io.pravega.common.util.BitConverter;
+import io.pravega.controller.server.SegmentHelper;
+import io.pravega.controller.server.rpc.auth.AuthHelper;
+import io.pravega.controller.store.index.ZKHostIndex;
+import io.pravega.controller.util.Config;
+import io.pravega.shared.NameUtils;
+import lombok.AccessLevel;
+import lombok.Getter;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.curator.framework.CuratorFramework;
+
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
+import java.util.concurrent.Executor;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.stream.Collectors;
+
+/**
+ * Pravega Tables stream metadata store.
+ */
+@Slf4j
+public class PravegaTablesStreamMetadataStore extends AbstractStreamMetadataStore {
+    static final String SEPARATOR = ".#.";
+    static final String SCOPES_TABLE = "Table" + SEPARATOR + "scopes";
+    static final String DELETED_STREAMS_TABLE = "Table" + SEPARATOR + "deletedStreams";
+    static final String COMPLETED_TRANSACTIONS_BATCHES_TABLE = "Table" + SEPARATOR + "completedTransactionsBatches";
+    static final String COMPLETED_TRANSACTIONS_BATCH_TABLE_FORMAT = "Table" + SEPARATOR + "completedTransactionsBatch-%d";
+
+    private static final String COMPLETED_TXN_GC_NAME = "completedTxnGC";
+
+    private final ZkInt96Counter counter;
+    private final ZKGarbageCollector completedTxnGC;
+    @VisibleForTesting
+    @Getter(AccessLevel.PACKAGE)
+    private final PravegaTablesStoreHelper storeHelper;
+    private final ZkOrderedStore orderer;
+
+    private final ScheduledExecutorService executor;
+    @VisibleForTesting
+    PravegaTablesStreamMetadataStore(SegmentHelper segmentHelper, CuratorFramework client, ScheduledExecutorService executor, AuthHelper authHelper) {
+        this(segmentHelper, client, executor, Duration.ofHours(Config.COMPLETED_TRANSACTION_TTL_IN_HOURS), authHelper);
+    }
+
+    @VisibleForTesting
+    PravegaTablesStreamMetadataStore(SegmentHelper segmentHelper, CuratorFramework curatorClient, ScheduledExecutorService executor, Duration gcPeriod, AuthHelper authHelper) {
+        super(new ZKHostIndex(curatorClient, "/hostTxnIndex", executor), new ZKHostIndex(curatorClient, "/hostRequestIndex", executor));
+        ZKStoreHelper zkStoreHelper = new ZKStoreHelper(curatorClient, executor);
+        this.orderer = new ZkOrderedStore("txnCommitOrderer", zkStoreHelper, executor);
```

Comment:
Yes, we have O(n) state per stream in zookeeper for two things --
1. stream reference used by Bucket Service for retention (and to be used for watermarking). 
This is basically `scopedStreamName` for each stream for which retention is enabled (and/or watermarking is enabled). 
2. references to all transactions that are currently marked for `commit` but the commit has not been processed yet.

Reply:
What prevented you from moving from zk to tables and is there a plan to eventually migrate it?

Reply:
Bucket service relies on "watch" on znodes to watch for any child addition or deletion. 
This ability to listen for changes to table is not there with our table implementation. So this will continue to be on zk.